### PR TITLE
2363/1

### PR DESCRIPTION
**Pull request by dixent**
## Номер
2363

## Номер задания
1

## Ссылка на видео с демо
https://vk.com/video?z=video194596469_456239080%2Fpl_cat_updates
https://vk.com/video194596469_456239081  видео с проверкой reek и rubocop

## Комментарии
Видео в vk, т.к. youtube забанил его.
Скидываю только все необходимые файлы(все с about_*.rb и triangle.rb).

### CODE REVIEW

**Comment by allomov** [comment]
Осторожно: коммит с видео.

**Comment by allomov** [comment]
добавь еще коммит, чтобы запустить собаку

File: 2363/1/triangle.rb [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+# This method smells of :reek:FeatureEnvy
+# This method smells of :reek:TooManyStatements
+def triangle(side_a, side_b, side_c)
+  sides = [side_a, side_b, side_c].sort
+  raise TriangleError if sides.any? { |side| side <= 0 } || ((sides[0] + sides[1]) <= sides[2])
+  sides.uniq!
+  hash = Hash.new(:scalene)
+  hash[1] = :equilateral
+  hash[2] = :isosceles
+  hash[sides.count]
```
Comment (allomov) [inline 2363/1/triangle.rb]: прочитай https://github.com/rubocop-hq/ruby-style-guide

**Comment by dixent** [comment]
Recheck

File: 2363/1/triangle.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+#   :scalene      if no sides are equal
+#
+# The tests for this method can be found in
+#   about_triangle_project.rb
+# and
+#   about_triangle_project_2.rb
+# This method smells of :reek:UtilityFunction
+# This method smells of :reek:FeatureEnvy
+# This method smells of :reek:TooManyStatements
+def triangle(*sides)
```
Comment (nplusp) [inline 2363/1/triangle.rb]: Hi. I'm a mentor from Rubizza. Alena asked to help a bit with reviews, so here am I :)

Interesting approach to solve the task, unusual indeed 👍

File: 2363/1/triangle.rb [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  sides.sort!
+  raise TriangleError if sides.any? { |side| side <= 0 } || ((sides[0] + sides[1]) <= sides[2])
```
Comment (nplusp) [inline 2363/1/triangle.rb]: Here you can replace `side <= 0` expression with `.negative?` method. And such way, you can transform the whole block `{ |side| side <= 0 }` to a `Symbol` (google `ruby method shorthands` 🌞)

You can ask "For what fucking sake I should do that?" and I'll patiently answer:
* first of all, it's shorter 
* secondly, it's actually a little bit faster! Yes, it's irrelevant here, but it's cool to know, amirite? 😉

```ruby
2.5.1 :053 > Benchmark.ips do |x|
2.5.1 :054 >     x.report('block with <=')  { sides.any? { |side| side <= 0 } }
2.5.1 :055?>     x.report('block with negative?')  { sides.any? { |side| side.negative? } }
2.5.1 :056?>     x.report('Symbol#to_proc') { sides.any?(&:negative?) }
2.5.1 :057?>     x.compare!
2.5.1 :058?> end

Warming up --------------------------------------
       block with <=   151.897k i/100ms
block with negative?   134.398k i/100ms
      Symbol#to_proc   149.107k i/100ms
Calculating -------------------------------------
      block with <=             2.636M (± 4.5%) i/s -     13.215M in   5.023704s
      block with negative?      2.158M (± 4.7%) i/s -     10.886M in   5.056460s
      Symbol#to_proc            2.668M (± 5.9%) i/s -     13.420M in   5.047230s

Comparison:
      Symbol#to_proc:        2667909.7 i/s
      block with <=:         2635739.8 i/s - same-ish: difference falls within error
      block with negative?:  2157561.1 i/s - 1.24x  slower
```

Comment (nplusp) [reply 2363/1/triangle.rb]: Ideally, you would like to split these validations and provide a message to these errors.

It's pretty easy thing to do:
```ruby
raise TriangleError, 'Side length cannot be less than a zero!' if ...
raise TriangleError, 'Sum of two lesser sides cannot be less that third side!' if ...
```

File: 2363/1/triangle.rb [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+  sides.uniq!
+  hash = Hash.new(:scalene).merge!(1 => :equilateral, 2 => :isosceles)
```
Comment (nplusp) [inline 2363/1/triangle.rb]: That's cool 👏🏻

File: 2363/1/triangle.rb (hunk 3 above)
Comment (dixent) [reply 2363/1/triangle.rb]: You are funny:) I will remake it, give me time

Comment (dixent) [reply 2363/1/triangle.rb]: I have some troubles. I reinstall my ubuntu yesterday. And now i can create commit in new PR. Will i create new PR and close this?

**Pull request closed without merging**